src/vos/vos_dtx.c

```diff
@@ -685,50 +685,64 @@ dtx_rec_release(struct vos_container *cont, struct vos_dtx_act_ent *dae, bool ab
 
 	if (dae->dae_records != NULL) {
 		D_ASSERT(DAE_REC_CNT(dae) > DTX_INLINE_REC_CNT);
+		D_ASSERT(!UMOFF_IS_NULL(dae_df->dae_rec_off));
 
-		for (i = DAE_REC_CNT(dae) - DTX_INLINE_REC_CNT - 1;
-		     i >= 0; i--) {
-			rc = do_dtx_rec_release(umm, cont, dae,
-						dae->dae_records[i], abort);
+		for (i = DAE_REC_CNT(dae) - DTX_INLINE_REC_CNT - 1; i >= 0; i--) {
+			rc = do_dtx_rec_release(umm, cont, dae, dae->dae_records[i], abort);
+			if (rc != 0)
+				return rc;
+		}
+
+		rc = umem_free(umm, dae_df->dae_rec_off);
+		if (rc != 0)
+			return rc;
+
+		if (keep_act) {
+			rc = umem_tx_add_ptr(umm, &dae_df->dae_rec_off, sizeof(dae_df->dae_rec_off));
 			if (rc != 0)
 				return rc;
+
+			dae_df->dae_rec_off = UMOFF_NULL;
 		}
 
+		DAE_REC_OFF(dae) = UMOFF_NULL;
+		dae->dae_records = NULL;
+
 		count = DTX_INLINE_REC_CNT;
 	} else {
 		count = DAE_REC_CNT(dae);
 	}
 
 	for (i = count - 1; i >= 0; i--) {
-		rc = do_dtx_rec_release(umm, cont, dae, DAE_REC_INLINE(dae)[i],
-					abort);
+		rc = do_dtx_rec_release(umm, cont, dae, DAE_REC_INLINE(dae)[i], abort);
 		if (rc != 0)
 			return rc;
-	}
 
-	if (!UMOFF_IS_NULL(dae_df->dae_rec_off)) {
-		rc = umem_free(umm, dae_df->dae_rec_off);
-		if (rc != 0)
-			return rc;
+		DAE_REC_INLINE(dae)[i] = UMOFF_NULL;
 	}
 
 	if (keep_act) {
+		if (count > 0) {
```

Comment:
Looks to me the count must be > 0, it's better to add a D_ASSERT(count > 0) before line 176?

Reply:
When re-commit former partial committed DTX, the count will be zero.